### PR TITLE
[pkg/translate/promremotewrite] downscale exponential histograms for prometheus

### LINKED ISSUE #17565: [pkg/translator/prometheusremotewrite] Implement Exponential Histogram downscaling
### Component(s)

pkg/translator/prometheusremotewrite

### Is your feature request related to a problem? Please describe.

When converting the Exponential Histogram to Prometheus Native Histogram its data points will be dropped if their scale is not in the `[-4, 8]` range.

### Describe the solution you'd like

Exponential Histogram data points should be downscaled to a scale accepted by Prometheus (range [-4,8]) if its scale is greater than 8. 

### Describe alternatives you've considered

Exponential Histogram data points can be dropped if their scales are greater than 8. 

### Additional context

See comments in https://github.com/open-telemetry/opentelemetry-specification/pull/3079

### PR DESCRIPTION
**Pull request by krajorama**
**Description:** 

Implement down-scaling of exponential histograms to Prometheus native histograms in Prometheus remote writer.
Configuration of down-scaling is TBD.

**Link to tracking Issue:**

Fixes: #17565 

**Testing:**

Unit tests.

**Documentation:**

TBD

### CODE REVIEW

**Comment by kovrus** [comment]
@krajorama please sing the CLA

File: pkg/translator/prometheusremotewrite/histograms.go [hunk 1]
```diff
@@ -123,34 +131,63 @@ func convertBucketsLayout(buckets pmetric.ExponentialHistogramDataPointBuckets)
...
 
-		// The offset is adjusted by 1 as described above.
-		bucketIdx := int32(i) + buckets.Offset() + 1
-		delta := bucketIdx - nextBucketIdx
-		if i == 0 || delta > 2 {
-			// We have to create a new span, either because we are
-			// at the very beginning, or because we have found a gap
+		gap := nextBucketIdx - bucketIdx - 1
+		if gap > 2 {
+			// We have to create a new span, either because we have found a gap
```
Comment (krajorama) [inline pkg/translator/prometheusremotewrite/histograms.go:155]:
```suggestion
			// We have to create a new span, because we have found a gap
```

**Comment by krajorama** [comment]
Current benchmark compared to original:
```
goos: linux
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/prometheusremotewrite
cpu: AMD Ryzen 7 4700U with Radeon Graphics         
                            │ bench-orig.txt │            bench-new.txt            │
                            │     sec/op     │   sec/op     vs base                │
ConvertBucketLayout/gap_0-8      17.30µ ± 1%   20.88µ ± 1%  +20.70% (p=0.000 n=10)
ConvertBucketLayout/gap_1-8      16.05µ ± 1%   19.09µ ± 0%  +18.89% (p=0.000 n=10)
ConvertBucketLayout/gap_2-8      15.61µ ± 1%   18.68µ ± 1%  +19.69% (p=0.000 n=10)
ConvertBucketLayout/gap_3-8      18.58µ ± 1%   21.51µ ± 2%  +15.77% (p=0.000 n=10)
geomean                          16.85µ        20.01µ       +18.75%
```

**Comment by krajorama** [comment]
After some optimizations I got this benchmark compared to baseline:
```
goos: linux
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/prometheusremotewrite
cpu: AMD Ryzen 7 4700U with Radeon Graphics         
                            │ bench-orig.txt │           bench-new3.txt           │
                            │     sec/op     │   sec/op     vs base               │
ConvertBucketLayout/gap_0-8      17.30µ ± 1%   18.18µ ± 1%  +5.07% (p=0.000 n=10)
ConvertBucketLayout/gap_1-8      16.05µ ± 1%   16.91µ ± 1%  +5.34% (p=0.000 n=10)
ConvertBucketLayout/gap_2-8      15.61µ ± 1%   16.94µ ± 1%  +8.55% (p=0.000 n=10)
ConvertBucketLayout/gap_3-8      18.58µ ± 1%   19.47µ ± 1%  +4.76% (p=0.000 n=10)
geomean                          16.85µ        17.84µ       +5.92%
```

Looking at the profile, half the time is spent in reallocating the deltas and probably spans slices. That could be something to look at , although for small numbers it's probably fine.

File: .chloggen/prometheusremotewrite-downscale-histogram.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: pkg/translator/prometheusremotewrite
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Downscale exponential histograms to fit prometheus native histograms.
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [1756]
```
Comment (kovrus) [inline .chloggen/prometheusremotewrite-downscale-histogram.yaml:15]:
```suggestion
issues: [17565]
```

**Review by kovrus — APPROVED** [review]
lgtm!

**Review by jpkrohling — APPROVED** [review]
I verified the tests, checking that the changes aren't breaking.

**Comment by jpkrohling** [comment]
CI is failing with an unrelated change, I'll rebase once the fix is merged.

**Review by jmacd — APPROVED** [review]

**Review by Aneurysm9 — COMMENTED** [review]
Couple nits and a question.  I'd approve but for the question about why apparently unrelated test changes seem to have been made.

File: pkg/translator/prometheusremotewrite/histograms.go [hunk 3]
```diff
@@ -123,34 +130,67 @@ func convertBucketsLayout(buckets pmetric.ExponentialHistogramDataPointBuckets)
 		prevCount = count
 	}
 
-	for i := 0; i < bucketCounts.Len(); i++ {
-		count := int64(bucketCounts.At(i))
+	// Let the compiler figure out that this is const during this function by
+	// moving it into a local variable.
+	bucketsCount := bucketCounts.Len()
```
[the remark this answers is not in the document]
Comment (krajorama) [reply pkg/translator/prometheusremotewrite/histograms.go]: thank you, renamed

File: pkg/translator/prometheusremotewrite/histograms_test.go [hunk 4]
```diff
@@ -141,7 +414,7 @@ func TestExponentialToNativeHistogram(t *testing.T) {
 				pt := pmetric.NewExponentialHistogramDataPoint()
 				pt.SetStartTimestamp(pcommon.NewTimestampFromTime(time.UnixMilli(100)))
 				pt.SetTimestamp(pcommon.NewTimestampFromTime(time.UnixMilli(500)))
-				pt.SetCount(2)
+				pt.SetCount(4)
```
[the remark this answers is not in the document]
Comment (krajorama) [reply pkg/translator/prometheusremotewrite/histograms_test.go:417]: I noticed that it was wrong when adding my own test case. I can put it into its own PR, but didn't seem worth it.

[the remark this answers is not in the document]
Comment (Aneurysm9) [reply pkg/translator/prometheusremotewrite/histograms_test.go:417]: Wrong how?  If changing these values don't cause the tests to fail are they even being used?  If not changing them causes the tests to fail then I'm suspicious of this change.

[the remark this answers is not in the document]
Comment (krajorama) [reply pkg/translator/prometheusremotewrite/histograms_test.go:417]: I'll add an assert on the count and fix the existing tests as needed.

[the remark this answers is not in the document]
Comment (krajorama) [reply pkg/translator/prometheusremotewrite/histograms_test.go:417]: I've added the asserts on both the input and output and sure enough the count calculation was off in the scale down case, very nice catch.

[the remark this answers is not in the document]
Comment (krajorama) [reply pkg/translator/prometheusremotewrite/histograms_test.go:417]: Wait a second, the count should not change due to scaling, it's the observation count which is not related to scale. Let me do this again.

[the remark this answers is not in the document]
Comment (krajorama) [reply pkg/translator/prometheusremotewrite/histograms_test.go:417]: Ok, actually there was no bug, I wrote the wrong assert. Anyway the upshot is that now those loose Count fields are properly validated for both the inputs and outputs of the test.

[the remark this answers is not in the document]
Comment (Aneurysm9) [reply pkg/translator/prometheusremotewrite/histograms_test.go:417]: So, to ensure I'm understanding correctly, the count fields were previously not validated and now they are and these two test cases had bad unchecked data?  Cool, I'm good with this then.

**Review by Aneurysm9 — APPROVED** [review]

**Comment by Dogacel** [comment]
Hoping this will resolve our issues with prometheus histograms 🙏 Putting the error we have seen here for visibility.

> Exporting failed. The error is not retryable. Dropping data. {"kind": "exporter", "data_type": "metrics", "name": "prometheusremotewrite", "error": "Permanent error: cannot convert exponential to native histogram. Scale must be <= 8 and >= -4, was 20;

**Pull request merged**